File: 백준/Gold/13913. 숨바꼭질 4/숨바꼭질 4.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Arrays;
import java.util.Queue;
import java.util.LinkedList;
import java.util.List;
import java.util.ArrayList;
import java.util.Stack;

public class Main {
	static int N,K,cur,pushItem;
	static int[] dist = new int[100001];
	static int[] vis = new int[100001];
	static Queue<Integer> q = new LinkedList<>();
	static Stack<Integer> st = new Stack<>();
	static List<Integer> li = new ArrayList<>();
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

	public static void main(String[] args) throws IOException {
		String[] sz = br.readLine().split(" ");
		N = Integer.parseInt(sz[0]);
		K = Integer.parseInt(sz[1]);
		Arrays.fill(dist,-1);
		if(N==K) {
			System.out.println(0);
			System.out.println(N);
			return;
		}
		bfs();
		System.out.println(dist[K]);
		pushItem=K;
		st.push(pushItem);
		while(pushItem!=N) {
			pushItem=vis[pushItem];
			st.push(pushItem);
		}
		while(!st.isEmpty()) System.out.print(st.pop()+" ");
	}
	
	static void bfs() {
		q.add(N);
		dist[N]=0;
		while(!q.isEmpty()) {
			cur = q.poll();
			int[] nxt = new int[] {cur-1, cur+1, 2*cur};
			for(int n : nxt) {
				if(n<0||n>100000) continue;
				if(dist[n]>=0||vis[n]>0) continue;
				dist[n]=dist[cur]+1;
				vis[n]=cur;
				q.add(n);
			}
 		}
	}
}

File: 백준/Gold/13913. 숨바꼭질 4/숨바꼭질 4.cc
#include <bits/stdc++.h>
using namespace std;
int N,K,dist[100001],pre[100001];
queue<int> q;
stack<int> st;
int main() {
    ios_base::sync_with_stdio(0); cin.tie(0);
    cin >> N >> K;
    if(N==K) {
        cout << 0 << '\n' << N << '\n';
        return 0;
    }
    fill(dist,dist+100001,-1);
    dist[N]=0;
    q.push(N);
    while(!q.empty()) {
        int cur = q.front(); q.pop();
        for(int nxt : {cur-1, cur+1, 2*cur}) {
            if(nxt<0||nxt>100000) continue;
            if(dist[nxt]!=-1) continue;
            dist[nxt]=dist[cur]+1;
            pre[nxt]=cur;
            q.push(nxt);
        }
    }
    st.push(K);
    int tmp=K;
    while(tmp!=N) {
        tmp=pre[tmp];
        st.push(tmp);
    }
    cout << dist[K] << '\n';
    while(!st.empty()) {
        int val=st.top(); st.pop();
        cout << val << ' ';
    }
    return 0;
}